### PR TITLE
File upload exercise support

### PR DESCRIPTION
**Pull request by terlan98**
This PR adds support for assessing file upload exercises

The following features, in particular, are added for file upload exercises:
- **Starting** a new assessment
- **Viewing** PDF, PNG, JPEG, DOCX, XLSX, CSV, and TXT files
- **Exporting** files with unsupported extensions to view externally
- Creating **general feedbacks**
- **Deleting and editing** feedbacks
- **Canceling** an assessment
- **Saving** an assessment
- **Submitting** an assessment
- Search for submissions & read-only mode
- The aforementioned operations for file upload exercises in **exams**

Here's what happens when a file extension is not supported:

https://github.com/ls1intum/Themis/assets/22798314/47cf7c03-225c-481d-8dcd-5962d30244ed

### CODE REVIEW

File: Themis/Views/Assessment/File Upload Exercise/FileUploadAssessmentView.swift [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    @ViewBuilder
+    private var fileRenderer: some View {
+        if let fileExtension = fileRendererVM.fileExtension {
+            if let localFileUrl = fileRendererVM.localFileURL {
+                FileRendererFactory.renderer(for: fileExtension, at: localFileUrl)
+            } else if let remoteFileUrl = fileRendererVM.remoteFileURL {
+                UnsupportedFileView(url: remoteFileUrl)
+            }
+        }
+    }
```
Comment (maximiliansoelch) [inline Themis/Views/Assessment/File Upload Exercise/FileUploadAssessmentView.swift]:
```suggestion
        if let fileExtension = fileRendererVM.fileExtension {
           let localFileUrl = fileRendererVM.localFileURL {
              
```

File: Themis/Views/Assessment/File Upload Exercise/FileUploadAssessmentView.swift [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    @Environment(\.presentationMode) private var presentationMode
+    @StateObject private var fileRendererVM = FileRendererViewModel()
+    @StateObject private var paneVM = PaneViewModel(mode: .rightOnly)
+    
+    private let didStartNextAssessment = NotificationCenter.default.publisher(for: NSNotification.Name.nextAssessmentStarted)
+    
+    var body: some View {
+        HStack(spacing: 0) {
+            if !fileRendererVM.isSetupComplete || fileRendererVM.isLoading {
+                VStack { // TODO: replace with a skeleton once the skeleton PR is merged
```
Comment (maximiliansoelch) [inline Themis/Views/Assessment/File Upload Exercise/FileUploadAssessmentView.swift]: Skeleton loading should be possible now

File: Themis/ViewModels/Assessment/File Upload/FileRendererViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    @Published var canDirectlyRenderFile = false
+    
+    var fileDownloadService = FileDownloadService()
+    
+    /// Sets this VM up based on the given submission
+    @MainActor
+    func setup(basedOn submission: BaseSubmission? = nil) async {
+        guard let fileUploadSubmission = submission as? FileUploadSubmission,
+              let baseUrl = UserSession.shared.institution?.baseURL?.absoluteString,
+              let filePath = fileUploadSubmission.filePath?.dropFirst(),
```
Comment (maximiliansoelch) [inline Themis/ViewModels/Assessment/File Upload/FileRendererViewModel.swift:28]: This line requires at least [artemis-ios-core-modules](https://github.com/ls1intum/artemis-ios-core-modules) version 3.4.0! 
This should be adapted in the project file.

File: Themis/Views/Assessment/File Upload Exercise/File Renderers/PDFRenderer.swift [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        return pdfView
+    }
+    
+    func updateUIView(_ uiView: PDFView, context: UIViewRepresentableContext<Self>) {}
+}
+
+struct PDFRenderer_Previews: PreviewProvider {
+    static var previews: some View {
+        // swiftlint:disable:next force_unwrapping
+        PDFRenderer(url: Bundle.main.url(forResource: "proposal", withExtension: "pdf")!)
```
Comment (maximiliansoelch) [inline Themis/Views/Assessment/File Upload Exercise/File Renderers/PDFRenderer.swift]: This causes preview crashes as this file is not part of the project.

**Review by maximiliansoelch — CHANGES_REQUESTED** [review]
Tested locally, found some issues:

1. When opening the second file upload submission, the console shows `FileRendererViewModel.setup():29 - Setup failed`
2. After some time, I receive a "Time Out Error". The problem statements are not loaded anymore and the currently logged-in user is not assigned to a course anymore. The user needs to restart in order to proceed with any assessment.


https://github.com/ls1intum/Themis/assets/9954674/ea715051-f0e2-4a33-bbce-02bf751112d1

**Comment by terlan98** [comment]
Thanks for the review! 🙏 

> 1. When opening the second file upload submission, the console shows `FileRendererViewModel.setup():29 - Setup failed`

This was because the 2nd submission was missing a file URL. I made some changes to handle such cases.

> 2. After some time, I receive a "Time Out Error". The problem statements are not loaded anymore and the currently logged-in user is not assigned to a course anymore. The user needs to restart in order to proceed with any assessment.

This was because of a broken branch deployed on the staging server. Should work fine with the develop branch.

File: Themis/Views/Assessment/File Upload Exercise/FileErrorIcon.swift [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+                    .padding(.top)
+                    .foregroundColor(backgroundColor)
+            }
+            .foregroundColor(.secondary)
+    }
+}
+
+#Preview {
+    FileErrorIcon()
+}
```
Comment (maximiliansoelch) [inline Themis/Views/Assessment/File Upload Exercise/FileErrorIcon.swift]: iOS17 is not yet supported on the build agent where we build Themis. Therefore, we need to use the old syntax so the build succeeds.

```suggestion
struct FileErrorIcon_Previews: PreviewProvider {
     static var previews: some View {
         FileErrorIcon()
     }
}
```

**Review by maximiliansoelch — CHANGES_REQUESTED** [review]

**Review by maximiliansoelch — COMMENTED** [review]
@terlan98 
For image files, we may also use a different option to show the image, as the current quicklook option allows the user to annotate the image. This might let the tutor think their annotations are also displayed to the student, which they are not.
Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2023-10-03 at 13 06 27

**Review by maximiliansoelch — APPROVED** [review]
Besides a few improvements, which can be done in a follow-up, the changes look good! Also tested them locally and they work as expected 👍

**Pull request merged**